Partial Assembly for Discontinuous Galerkin

YohannDudouit opened the pull request:
This PR proposes an implementation of Partial Assembly for Discontinuous Galerkin formulations for Gauss-Lobatto (or Bernstein) basis functions. The target of this PR is to make `ex9` and `ex9p` work with partial assembly.

__Progress:__
- [x] 2D `ex9` working with Gauss-Lobatto basis functions
- [x] 2D `ex9p` working with Gauss-Lobatto basis functions
- [x] 3D `ex9` working with Gauss-Lobatto basis functions
- [x] 3D `ex9p` working with Gauss-Lobatto basis functions
- [x] Separate the treatment of true boundary faces and interior faces
- [x] Test in Device mode
- [x] Optimization of the kernels

__New classes:__
- `H1FaceRestriction`: Provides dofs for each face of a H1 finite element space. (Only for Gauss-Lobatto)
- `L2FaceRestriction`: Provides dofs for each trace element on each face of a L2 finite element space. (Only for Gauss-Lobatto)
- `ParL2FaceRestriction`: Same as `L2FaceRestriction` but for `ParFESpace`.
- `FaceQuadratureInterpolator`:  Compute values, jacobians, det(J), normals on the faces.

__Extension of existing classes:__
- PA implementation for Convection Integrator.
- PA implementation for DGTrace Integrator.


- [x] Code builds.
- [x] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [ ] Had a new optional library been added? (*Make sure the external library is licensed under LGPL, not GPL!*)
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [x] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that are generated from the source but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New capability:
   - [ ] All significant new classes, methods and functions have Doxygen-style documentation in source comments.
   - [x] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [x] Run the unit tests and make sure they all pass `make unittest`.
- [ ] (LLNL only) Clone the `tests` repository and run the following tests, see `mfem/tests/README.md`:
   - [ ] `compilers`
   - [ ] `memcheck`
   - [ ] `documentation`
- [ ] (LLNL only) After merging:
   - [ ] Regenerate `README.html` files from companion documentation pull requests.
   - [ ] Update the `baseline` and `compiler` tests, add new tests if necessary.
   - [ ] Consider updating the script `mfem/tests/sample-runs` (`sample-runs-serial` and `sample-runs-parallel`).

<!--GHEX{"id":1207,"author":"YohannDudouit","editor":"tzanio","reviewers":["artv3","pazner","v-dobrev","camierjs"],"assignment":"2020-01-14","approval":"2020-03-02T22:50:17.626Z","merge":"2020-03-05T06:13:11.700Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1207](https://github.com/mfem/mfem/pull/1207) | @YohannDudouit | @tzanio | @artv3 + @pazner + @v-dobrev + @camierjs | 01/14/20 | 03/02/20 | 03/04/20 | |
<!--ELBATXEHG-->

YohannDudouit commented:
@tzanio @v-dobrev Everything seem to be working. I'm starting the final phase of cleaning the code, improving the documentation, more thorough testing, and benchmarking.

tzanio commented:
Please address Appveyor errors.

artv3 commented on examples/ex9p.cpp:
```diff
@@ -198,11 +213,13 @@ int main(int argc, char *argv[])
    // 7. Define the parallel discontinuous DG finite element space on the
    //    parallel refined mesh of the given polynomial order.
-   DG_FECollection fec(order, dim);
+   DG_FECollection fec(order, dim, BasisType::GaussLobatto);
```
yes!!!!

pazner commented on fem/pfespace.cpp:
```diff
@@ -18,6 +18,7 @@
 #include "../general/sort_pairs.hpp"
 #include "../mesh/mesh_headers.hpp"
 #include "../general/binaryio.hpp"
+#include "fespace.hpp"
```
This include is not necessary since `fespace.hpp` is included by `pfespace.hpp`.

pazner commented on fem/pfespace.cpp:
```diff
@@ -3245,6 +3422,252 @@ void DeviceConformingProlongationOperator::MultTranspose(const Vector &x,
+   const int nd = dof;
+   const int vd = vdim;
+   const bool t = byvdim;
+   const int treshold = ndofs;
```
Typo:
```suggestion
   const int threshold = ndofs;
```

pazner commented on mesh/mesh.cpp:
```diff
@@ -3850,7 +3871,7 @@ void Mesh::SetNodalFESpace(FiniteElementSpace *nfes)
 {
    if (Nodes) { return; }
-   SetCurvature(1, false, -1, Ordering::byVDIM);
+   SetCurvature(1, false, -1, Ordering::byNODES);
```
Is this change required?

pazner commented on mesh/mesh.cpp:
```diff
@@ -10069,9 +10088,95 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
+   }
+}
+
+static void GetSigns(const FiniteElementSpace &fes, const FaceType type,
```
Documentation here would be helpful!

pazner commented on fem/bilininteg_convection.cpp:
```diff
@@ -0,0 +1,869 @@
+         case 0x66: return SmemPAConvectionApply3D<6,6>(NE,B,G,Bt,Gt,op,x,y);
+         case 0x77: return SmemPAConvectionApply3D<7,7>(NE,B,G,Bt,Gt,op,x,y);
+         case 0x88: return SmemPAConvectionApply3D<8,8>(NE,B,G,Bt,Gt,op,x,y);
+         default:   return PAConvectionApply3D(NE,B,G,Bt,Gt,op,x,y,D1D,Q1D);
```
Need to decide what to do here regarding #1226.

pazner commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,950 @@
+         case 0x89: return SmemPADGTraceApply3D<8,9,1>(NF,B,Bt,op,x,y);
+         default:   return PADGTraceApply3D(NF,B,Bt,op,x,y,D1D,Q1D);
+      }
+   }
```
Need to decide what to do here regarding #1226.

pazner commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,950 @@
+         case 0x67: return PADGTraceApplyTranspose3D<6,7>(NF,B,Bt,op,x,y);
+         case 0x78: return PADGTraceApplyTranspose3D<7,8>(NF,B,Bt,op,x,y);
+         case 0x89: return PADGTraceApplyTranspose3D<8,9>(NF,B,Bt,op,x,y);
+         default: return PADGTraceApplyTranspose3D(NF,B,Bt,op,x,y,D1D,Q1D);
```
As above, cf. #1226.

pazner commented on fem/fespace.cpp:
```diff
@@ -855,6 +870,181 @@ const Operator *FiniteElementSpace::GetElementRestriction(
+}
+
+const Operator *FiniteElementSpace::GetSingleValuedFaceRestriction(
+   ElementDofOrdering e_ordering, FaceType type) const
```
Can some code duplication be reduced here by combining with `GetFaceRestriction`?

YohannDudouit replied:
I simplified the code and removed the `GetSingleValuedFaceRestriction` function.

pazner commented on fem/fespace.cpp:
```diff
@@ -2882,238 +3136,1260 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
+   // {
+   //    j = size1d-1-j;
+   // }
+   // rotate on lex ordering
```
Delete commented out code here and below?

pazner commented on fem/fespace.cpp:
```diff
@@ -2882,238 +3136,1260 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
+      case 3:
+         return PermuteFace3D(face_id1, face_id2, orientation, size1d, index);
+      default:
+         mfem_error("Incorrect dimension.");
```
```suggestion
         mfem_error("Unsupported dimension.");
```

pazner commented on fem/fespace.cpp:
```diff
@@ -2882,238 +3136,1260 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
+            //    {
+            //       der(q,c,f) = D[c];
+            //    }
+            // }
```
Remove?

pazner commented on fem/fespace.cpp:
```diff
@@ -2882,238 +3136,1260 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
+         //       der(q,c,1,e) = D[c+VDIM*1];
+         //       der(q,c,2,e) = D[c+VDIM*2];
+         //    }
+         // }
```
Remove?

pazner commented on fem/fespace.hpp:
```diff
@@ -1010,6 +1098,115 @@ class QuadratureInterpolator
+class FaceQuadratureInterpolator
+{
+private:
+   FaceType type;
```
I get warnings about this being unused when compiling with clang. However, if I delete this line I get compilation errors. Should look into this.

YohannDudouit commented:
@tzanio @adrienbernede Travis complains because I touched the `mesh.cpp` file, I think this will require manual intervention.

adrienbernede commented:
@YohannDudouit @tzanio 
Well, it seems I failed to catch it during review, but the optional job should be in the "allow_failures" section (instead of "include"). I can push a fix tomorrow, test it with your branch and we could have it (the fix) merged into master.

tzanio commented:
> @tzanio @adrienbernede Travis complains because I touched the mesh.cpp file, I think this will require manual intervention.

This is actually intentional. We can adjust the limits for size, line changes etc. or filter some of the known big files in the repo. But we can also leave it as is so we double-check before the merge is executed (the merge button is red and requires administrative privilege).

YohannDudouit commented on examples/ex9p.cpp:
```diff
@@ -198,11 +213,13 @@ int main(int argc, char *argv[])
-   DG_FECollection fec(order, dim);
+   DG_FECollection fec(order, dim, BasisType::GaussLobatto);
    ParFiniteElementSpace *fes = new ParFiniteElementSpace(pmesh, &fec);
+   fes->ExchangeFaceNbrData();
```
The one for the `pmesh` was not necessary, I removed it, however the one for `fes` is necessary (SegFault otherwise).
@v-dobrev Is it okay to add `ExchangeFaceNbrData()` in the constructor of `ParFiniteElementSpace`?

camierjs replied:
I agree that this should be triggered somewhere internally.

YohannDudouit commented on fem/fespace.hpp:
```diff
@@ -1010,6 +1101,115 @@ class QuadratureInterpolator
+
+   static const int MAX_NQ3D = 1000;
+   static const int MAX_ND3D = 1000;
+   static const int MAX_VDIM3D = 3;
```
This is also in the `QuadratureInterpolator` class, I agree that this duplication is bad, and these constants should probably be defined somewhere else, however I'm not sure defining these constants as globals is cleaner, do you have a suggestion?

camierjs replied:
Not a fan too, but no real idea besides waiting for some `JIT` capabilities.

artv3 commented:
@YohannDudouit, thanks for working on this! I'll give it a spin this week 👍

camierjs commented on fem/CMakeLists.txt:
```diff
@@ -19,6 +19,8 @@ set(SRCS
   bilininteg_mass.cpp
   bilininteg_vecdiffusion.cpp
   bilininteg_vecmass.cpp
+  bilininteg_convection.cpp
```
How does this PR combines with the previous `convection` one?

artv3 replied:
I think these kernels are more optimized, feel free replace the previous ones. However; I would like if we can maintain the same capabilities in particular ability to handle variable coefficients.

YohannDudouit replied:
There is a big merge conflict, I'll solve it today.

YohannDudouit replied:
The kernels are surprisingly similar, there is very little difference between the your kernels and my unoptimized kernels. I was having `out of memory` problems in 3D though, were you facing similar problems?

YohannDudouit replied:
If that's okay with you I'll remove the case `Q==NULL` where you set the velocity field to `(1,1,1)`, that doesn't really make sense to me. That makes sense for a scalar field to set it to `1`, but not for a vectorial field.

camierjs commented on fem/bilininteg.cpp:
```diff
@@ -22,7 +22,7 @@ namespace mfem
 void BilinearFormIntegrator::AssemblePA(const FiniteElementSpace&)
 {
-   mfem_error ("BilinearFormIntegrator::AssemblePA(...)\n"
+   mfem_error ("BilinearFormIntegrator::AssemblePA (...)\n"
```
Shouldn't be touched I think.

camierjs commented on fem/fespace.cpp:
```diff
@@ -2881,238 +3002,1215 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
    });
 }
 
-
```
There is a lot of changes in this already huge file.
Some of the them collides with existing code, even GitHub is confused.
Open question: Could we start splitting this `Face` additions to a new files?

YohannDudouit replied:
@tzanio @v-dobrev Is it okay if I move all the code related to restrictions to `fem/restriction.hpp` and 'fem/restriction.cpp`?

tzanio replied:
If it is a separate class, yes. Just avoid having implementations of the same class in multiple files.

v-dobrev replied:
Sounds good to me.

YohannDudouit replied:
done.

camierjs commented on fem/pgridfunc.cpp:
```diff
@@ -225,19 +235,21 @@ void ParGridFunction::ExchangeFaceNbrData()
 
-   const double *h_data = this->HostRead();
-   for (int i = 0; i < send_data.Size(); i++)
+   auto d_data = this->Read();
```
Would there be a way to use direct pointers when `GetGPUAwareMPI` is `true`?

YohannDudouit replied:
I added the `GPUAwareMPI` code.

camierjs commented on general/cuda.hpp:
```diff
@@ -48,17 +48,40 @@
 
 // Define the MFEM inner threading macros
 #if defined(MFEM_USE_CUDA) && defined(__CUDA_ARCH__)
+#if __CUDA_ARCH__ < 600
```
We should have a way to get rid of these atomics.

It's okay to have some kernels using them and being documented, but inside `MFEM`, we should have a way to fold back to a version not using them.

camierjs replied:
We should get rid of this `MFEM_ATOMIC_ADD` code now that is not been used anymore.

artv3 commented on fem/pfespace.cpp:
```diff
@@ -3252,6 +3281,252 @@ void DeviceConformingProlongationOperator::MultTranspose(const Vector &x,
+         for (int c = 0; c < vd; ++c)
+         {
+            MFEM_ATOMIC_ADD(d_y(t?c:idx1,t?idx1:c), d_x(i % nd, c, 0, i / nd));
+            if (idx2>-1 && idx2<threshold) { MFEM_ATOMIC_ADD(d_y(t?c:idx2,t?idx2:c), d_x(i % nd, c, 1, i / nd)); }
```
I see we are using atomics here. Could this be avoided easily?

YohannDudouit replied:
Not easily. But this can be avoided, using a quite different algorithm.

YohannDudouit replied:
It's easier than I thought, next thing on my TODO list.

YohannDudouit replied:
done.

artv3 commented:
One thing that would also be good to have is a unit test comparing the convection + trace operator action of PA vs FA.

YohannDudouit commented:
@artv3 This is what `ex9pa` does, I need to transform it into a test.

pazner commented on fem/fespace.cpp:
```diff
@@ -2881,238 +3002,1215 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
+               {
+                  faceMap[i] = (dof1d-1)*dof1d - i*dof1d;
+               }
+               break;
```
This should be made lexicographic to match with the 3D implementation

YohannDudouit replied:
done.

artv3 commented:
> Have you noticed that the CUDA runs are slower than the ones on a single CPU core?

Which version did you try ex9 or ex9p? Do we know what might be the bottleneck?

YohannDudouit commented:
@artv3 To be more accurate, we checked with @camierjs , `ex9` is slow for small number of unknowns, for bigger problems GPU is much faster than CPU. The cause are `L2Elementrestriction`, and the CG algorithm which is done globally instead of locally, both are not additions of this PR.

artv3 commented on fem/pgridfunc.cpp:
```diff
@@ -242,18 +242,19 @@ void ParGridFunction::ExchangeFaceNbrData()
+   bool mpi_gpu_aware = Device::GetGPUAwareMPI();
+   auto send_data_ptr = mpi_gpu_aware ? send_data.Read() : send_data.HostRead();
+   auto face_nbr_data_ptr = mpi_gpu_aware ? face_nbr_data.Write() : face_nbr_data.HostWrite();
    for (int fn = 0; fn < num_face_nbrs; fn++)
```
Nice! Can you let me know this improves performance? I use this function quite a bit.

YohannDudouit replied:
@camierjs Do you have any information on this?

camierjs replied:
I'm using it locally, but haven't re-tested it on some large run.

tzanio commented:
There have been a lot of changes to the PR recently, so I am extending the deadline by a week to give @artv3 + @pazner + @v-dobrev + @camierjs more time to review.

YohannDudouit commented:
@pazner @camierjs @artv3 @v-dobrev  I have addressed all the comments from your previous reviews. You can restart your review.
@camierjs I've rewritten `L2ElementRestriction` and added an optimized version of `PADGTraceApplyTranspose` with good impact on performance.
All the code related to `ElementRestriction` and `FaceRestriction` is now in `restriction.hpp/cpp`.

pazner commented on examples/ex9.cpp:
```diff
@@ -237,7 +251,7 @@ int main(int argc, char *argv[])
 
```
```suggestion
   if (pa && mesh.GetNodes() != NULL)
   {
      mesh.SetCurvature(mesh.GetNodalFESpace()->GetOrder(0), true);
   }
```

Something like this should be added to make sure ex9 works with e.g. star-q3

YohannDudouit replied:
done.

pazner commented on examples/ex9.cpp:
```diff
@@ -403,13 +425,23 @@ int main(int argc, char *argv[])
+      dg_solver = new DG_Solver(M.SpMat(), K.SpMat(), fes);
+      M_solver.SetOperator(M.SpMat());
+   }
 
```
```suggestion
    M_solver.SetPreconditioner(*M_prec);
```
Oops

pazner commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,1092 @@
+   {
+      for (int q = 0; q < Q1D; ++q)
+      {
+         const double r = const_r ? R(0,0) : R(q,f);
```
Compiler warning here

pazner commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,1092 @@
+      {
+         for (int q2 = 0; q2 < Q1D; ++q2)
+         {
+            const double r = const_r ? R(0,0,0) : R(q1,q2,f);
```
Compiler warning here

artv3 commented on fem/bilininteg_convection.cpp:
```diff
@@ -163,292 +147,667 @@ void PAConvectionApply2D(const int NE,
+      {
+         MFEM_FOREACH_THREAD(dx,x,D1D)
+         {
+            u[tidz][dy][dx] = x(dx,dy,e);
```
Looking at this, it looks like every thread along the z-dimension is reading the same (dx,dy) entry. I think we are trying to stack along the z-dimension, but I think your missing the z-offset when accessing x(....) right?
To be more clear, I think thread (0,0,0) is reading the same data as (0,0,1) where I'm enumerating my threads (x,y,z).

YohannDudouit replied:
I'll add a comment to clarify this, but `e` is different along the z-dimension, i.e. `e=e+tidz` you can check the implementation of `MFEM_FORALL_2D`.

YohannDudouit commented on fem/restriction.cpp:
```diff
@@ -0,0 +1,1351 @@
+ParL2FaceRestriction::ParL2FaceRestriction(const ParFiniteElementSpace &fes,
+                                           ElementDofOrdering e_ordering,
+                                           FaceType type,
+                                           L2FaceValues m)
```
I completely agree with this, unfortunately I don't think it is possible because most of the `Par` objects don't even exist in serial. Also, the code is not really duplicated, there is differences.

YohannDudouit commented on fem/restriction.hpp:
```diff
@@ -0,0 +1,151 @@
+
+/** An enum type to specify if only e1 value is requested (Single) or both
+    e1 and e2 (Double). */
+enum class L2FaceValues : bool {Single, Double};
```
I picked the first one: `SingleValued`, `DoubleValued`.

YohannDudouit commented:
@artv3 @camierjs @v-dobrev We're back to a stable state after @pazner's last review.

pazner commented on examples/ex9p.cpp:
```diff
@@ -218,6 +229,9 @@ int main(int argc, char *argv[])
    }
 
+   Device device(device_config);
+   device.Print();
```
```suggestion
   if (myid == 0) { device.Print(); }
```

YohannDudouit replied:
done.

pazner commented on fem/restriction.hpp:
```diff
@@ -0,0 +1,153 @@
+   const L2FaceValues m;
+   const int nfdofs;
+   Array<int> scatter_indices2;
+   Array<int> scatter_indices1;
```
Ordering here causes a compilation warning.

YohannDudouit replied:
done.

artv3 commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,1092 @@
+   auto Bt = Reshape(bt.Read(), D1D, Q1D);
+   auto op = Reshape(_op.Read(), Q1D, 2, 2, NF);
+   auto x = Reshape(_x.Read(), D1D, VDIM, 2, NF);
+   auto y = Reshape(_y.ReadWrite(), D1D, VDIM, 2, NF);
```
what does the 2 in line 268 correspond to?

YohannDudouit replied:
`Elem1` or `Elem2` on the face.

artv3 replied:
How do you account for an element being on the boundary?

YohannDudouit replied:
Interior and boundary faces are treated independently since you can't assume the same Integrators.

YohannDudouit replied:
You specify to `FaceRestriction` if you're treating boundary or interior faces.

camierjs commented on fem/fespace.hpp:
```diff
@@ -1012,6 +1030,113 @@ class QuadratureInterpolator
+    The target quadrature points in the elements can be described either by an
+    IntegrationRule (all mesh elements must be of the same type in this case) or
+    by a QuadratureSpace. */
+class FaceQuadratureInterpolator
```
It would be also good to start to avoid putting things more in `fespace.*`, shouldn't we open a new file for the `QuadratureInterpolator`?

YohannDudouit replied:
I'm okay with it, @v-dobrev @tzanio ?

tzanio replied:
I'm fine with it, as long as it is an independent class that makes sense on its own.

v-dobrev replied:
Moving the `*Interpolator` classes to a new file sounds good to me -- it will probably speed up the build a little. :smile:

YohannDudouit replied:
done.

camierjs commented on mesh/mesh.cpp:
```diff
@@ -777,13 +777,39 @@ const GeometricFactors* Mesh::GetGeometricFactors(const IntegrationRule& ir,
    return gf;
 }
 
+const FaceGeometricFactors* Mesh::GetFaceGeometricFactors(
```
Again, adding new `face` features in `mesh` could be avoided if we can come with new files.

YohannDudouit replied:
I'm not sure we can easily get rid of this function since the `FaceGeometricFactors` are stored in the mesh similarly to the `GeometricFactors`.

camierjs commented on tests/unit/fem/test_pa_kernels.cpp:
```diff
@@ -353,97 +353,113 @@ TEST_CASE("PA Vector Diffusion", "[PartialAssembly], [VectorPA]")
 }
 
-//Basic unit test for convection
-TEST_CASE("PA Convection")
```
There is a lot of changes again in this file initiated by @artv3 - we want to make sure both sides are good with these.

YohannDudouit replied:
The idea is just to extend the "PA Convection" tests to DG, what was done by @artv3 is unchanged.

artv3 replied:
Looks good to me 👍

YohannDudouit commented on fem/bilininteg_convection.cpp:
```diff
@@ -20,139 +20,123 @@ namespace mfem
 
 // PA Convection Integrator
```
As mentioned during the merge the algorithm are almost identical for non optimized versions, they just used a different data layout for the quadrature data (also slightly better for memory accesses), that's why I kept my implementations.

YohannDudouit replied:
@artv3 also suggested to keep my implementations.(https://github.com/mfem/mfem/pull/1207#discussion_r372606427)

vladotomov commented on tests/unit/fem/test_face_permutation.cpp:
```diff
@@ -0,0 +1,231 @@
+
+      for (int i=0; i<ndof_face; ++i)
+      {
+         REQUIRE(face_values[i] == face_values(i + ndof_face));
```
Brackets [ ] ( )

YohannDudouit replied:
I'm not sure I understand why it compiled with `[]` since `vector` doesn't seem to have this operator. I made the change.

pazner replied:
I guess it's doing an implicit cast to `double *` first but that way you don't get bounds checking in debug mode.

vladotomov commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,1092 @@
+{
+// PA DG Trace Integrator
+
+// PA DG Trace Assemble 2D kernel for constant velocity
```
Old comment?

YohannDudouit replied:
removed.

artv3 commented on mesh/pmesh.hpp:
```diff
@@ -121,6 +121,9 @@ class ParMesh : public Mesh
                               const STable3D &old_faces,
                               Array<int> *f2qf);
 
+   virtual void SetCurvature(int order, bool discont = false, int space_dim = -1,
```
Can we keep this public? This is useful for applications.

artv3 commented on fem/bilininteg_convection.cpp:
```diff
@@ -163,292 +147,668 @@ void PAConvectionApply2D(const int NE,
+   ElementTransformation &Trans = *fes.GetElementTransformation(0);
+   const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el, Trans);
+   const int dims = el.GetDim();
+   const int symmDims = dims; // 1x1: 1, 2x2: 3, 3x3: 6
```
Symmdims? I think the comment should be along the lines of ``// 1x1: 1, 2x1: 2, 3x1: 3``.

YohannDudouit replied:
I removed the comment.

tzanio commented:
@v-dobrev + @camierjs, the review on this PR is due this week

camierjs commented:
I have issues testing, running into:
```mpicxx  -g -Wall -std=c++11 -I../.. -I../../../hypre/src/hypre/include -c display-basis.cpp
mpicxx -g -Wall -std=c++11 -I../.. -I../../../hypre/src/hypre/include -o display-basis display-basis.o -L../../miniapps/common -lmfem-common  -L../.. -lmfem -L../../../hypre/src/hypre/lib -lHYPRE  -L../../../metis-4.0 -lmetis -lrt
display-basis.o: In function `update_basis(std::vector<mfem::socketstream*, std::allocator<mfem::socketstream*> >&, VisWinLayout const&, mfem::Element::Type, char, int, int, Deformation::DefType, DeformationData const&, bool, int&)':
/home/camier1/home/mfem/yohann-dg/miniapps/tools/display-basis.cpp:707: undefined reference to `mfem::common::ElementMeshStream::ElementMeshStream(mfem::Element::Type)'
collect2: error: ld returned 1 exit status
make[1]: *** [display-basis] Error 1
make[1]: Leaving directory `/home/camier1/home/mfem/yohann-dg/miniapps/tools'
make: *** [miniapps/tools] Error 2
```

YohannDudouit commented:
@camierjs All tests pass locally and for Travis, are you sure the problem doesn't come from your side?

camierjs commented:
My bad, I was hitting some `MINIAPP_USE_COMMON` issue that is fixed in #1104.

artv3 commented:
As a heads up there will be merge conflicts with PR: https://github.com/mfem/mfem/pull/1208. I tried consolidating the work earlier today but ran into conflicts.

artv3 commented on fem/pgridfunc.cpp:
```diff
@@ -225,23 +235,27 @@ void ParGridFunction::ExchangeFaceNbrData()
 
-   const double *h_data = this->HostRead();
-   for (int i = 0; i < send_data.Size(); i++)
+   auto d_data = this->Read();
```
This breaks my application. **This** could be a alias, and the memory manager would throws an error about registering an alias.

artv3 replied:
@v-dobrev , do you know what a good fix for this might be? I tried doing something like:
``` 
Vector mpi_data; 
mpi_data.MakeRef(*this,0);
```

but surprisingly it still gave me the same error.

artv3 replied:
Actually, I'm wondering if this might require code changes on our end. I'll keep y'all in the loop.

YohannDudouit replied:
I close this for now since it seems to be on your side.

tzanio commented:
Adding a few extra days for approval...

camierjs commented:
Some of the sample runs fail:
``` 
Verification failed: (fes.GetMesh()->Conforming()) is false:
 --> Non-conforming meshes not yet supported with partial assembly.
 ... in function: mfem::H1FaceRestriction::H1FaceRestriction(const mfem::FiniteElementSpace&, mfem::ElementDofOrdering, mfem::FaceType)
 ... in file: fem/restriction.cpp:328
```
Can be reproduced with `./ex6 -pa`, or on Lassen with:
`lalloc 1 -q pdebug -G ceed`
`make distclean && make pcuda -j && make all -j`
`./config/sample-runs.sh -b -p -dev -j 32`

Not sure why are we going down to the face restriction there.

YohannDudouit commented:
@camierjs I fixed the `Update()` function, all the sample runs now pass.

v-dobrev commented on examples/ex9.cpp:
```diff
@@ -263,11 +282,14 @@ int main(int argc, char *argv[])
+   {
+      m.Finalize();
+      k.Finalize(skip_zeros);
+   }
```
I'd move the check for assembly level inside the `Finalize()` method -- keeps it simple in the example.

v-dobrev commented on examples/ex9.cpp:
```diff
@@ -345,7 +367,7 @@ int main(int argc, char *argv[])
    //    right-hand side, and perform time-integration (looping over the time
    //    iterations, ti, with a time-step dt).
-   FE_Evolution adv(m.SpMat(), k.SpMat(), b, fes);
+   FE_Evolution adv(m, k, b, fes, pa);
```
The `pa` parameter can be obtained from the `BilinearForm`s, no need to pass it separately.

v-dobrev replied:
Actually, `fes` can also be obtained from the `BilinearForm`s.

v-dobrev commented on examples/ex9p.cpp:
```diff
@@ -407,7 +437,7 @@ int main(int argc, char *argv[])
    //     right-hand side, and perform time-integration (looping over the time
    //     iterations, ti, with a time-step dt).
-   FE_Evolution adv(*M, *K, *B, *fes);
+   FE_Evolution adv(*M, *K, *B, *fes, pa);
```
To make the code more similar to the serial version, it seems more natural to pass the `ParBilinearForm`s as parameters here.

YohannDudouit replied:
Do you mean both `m` and `k`? That's not very elegant because it's redundant I think.

YohannDudouit replied:
I'll try to initialize `M` and `K` inside the `FE_Evolution`

YohannDudouit replied:
I changed the interface of `FE_Evolution` to be similar to `ex9`, I'm not completely convinced about the elegance of the code, tell me if you like it.

v-dobrev replied:
I think you can use `OperatorPtr` in the `FE_Evolution` class to simplify the code a little but I'll leave that up to you.

v-dobrev commented on fem/bilinearform_ext.hpp:
```diff
@@ -102,11 +102,16 @@ class PABilinearFormExtension : public BilinearFormExtension
    const FiniteElementSpace *trialFes, *testFes; // Not owned
    mutable Vector localX, localY;
+   mutable Vector faceIntX, faceIntY;
+   mutable Vector faceBoundX, faceBoundY;
```
In other places of the code "boundary" is typically abbreviated as "bdr". It can changed later maybe.

v-dobrev commented on fem/bilinearform_ext.cpp:
```diff
@@ -42,24 +42,76 @@ PABilinearFormExtension::PABilinearFormExtension(BilinearForm *form)
+   {
+      int_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC, FaceType::Interior);
+      if (int_face_restrict_lex)
```
This check seems redundant -- unlike the element restriction case, the face restrictions can never be the identity operator, right?

v-dobrev commented on fem/bilinearform_ext.cpp:
```diff
@@ -100,6 +152,30 @@ void PABilinearFormExtension::Update()
    }
+
+   const int iFISz = a->GetFBFI()->Size();
+   if (int_face_restrict_lex && iFISz>0)
```
Maybe `SetupRestrictionOperators()` can be reused here? Actually that will be done by `Assemble()` (which should be called after a call to this method). It seems we can  just reset the restriction pointers to `nullptr` here.

YohannDudouit replied:
Good suggestion.

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -2436,6 +2473,21 @@ class DGTraceIntegrator : public BilinearFormIntegrator
+   static const IntegrationRule &GetRule(Geometry::Type geom, int order,
+                                         FaceElementTransformations &T);
+
+   void SetupPA(const FiniteElementSpace &fes, FaceType type);
```
It seems this method is neither defined nor used anywhere.

YohannDudouit replied:
it is defined [here](https://github.com/mfem/mfem/pull/1207/files/45a299b04aa0febbd07a8b6aef928262c4db1035#diff-71b826a0bdf2bdffe818868995499256R134) in ` fem/bilininteg_dgtrace.cpp`, and used in the `AssemblePAInteriorFaces` and `AssemblePABoundaryFaces` methods.

v-dobrev replied:
Oh OK, I guess I did not look in that file.

If it is used only internally, how about we make it protected?

YohannDudouit replied:
I made it `private`.

v-dobrev commented on fem/fe_coll.cpp:
```diff
@@ -1825,8 +1825,15 @@ L2_FECollection::L2_FECollection(const int p, const int dim, const int btype,
+      else
+      {
+         Tr_Elements[Geometry::SEGMENT] = new L2_SegmentElement(p, btype);
+      }
```
I think this is for simplicity -- every time we use the trace elements we can assume they are nodal, e.g. in `Mesh::GetBdrElementTransformation`.

v-dobrev commented on fem/fespace.hpp:
```diff
@@ -128,8 +135,36 @@ class FiniteElementSpace
    /// The element restriction operators, see GetElementRestriction().
    mutable OperatorHandle L2E_nat, L2E_lex;
+   /// The face restriction operators, see GetFaceRestriction().
+   using key_face = std::tuple<bool, ElementDofOrdering, FaceType, L2FaceValues>;
```
The code may be a bit clearer (for people like me -- I had to look up what `std::get` does) if `key_face` is a simple struct

v-dobrev commented on fem/fespace.hpp:
```diff
@@ -128,8 +135,36 @@ class FiniteElementSpace
+                   std::get<3>(v0) == std::get<3>(v1)
+                );
+      }
+   };
```
Isn't this the same as the standard `operator==()` for `tuple`s?

YohannDudouit replied:
You're correct.

v-dobrev commented on fem/fespace.hpp:
```diff
@@ -338,6 +379,14 @@ class FiniteElementSpace
+      const IntegrationRule &ir, FaceType type) const;
+
+   const FaceQuadratureInterpolator *GetFaceQuadratureInterpolator(
+      const QuadratureSpace &qs, FaceType type) const;
```
I don't think `QuadratureSpace` is the right type here. The method that is used by the returned `FaceQuadratureInterpolator` (in its `Mult` method) is `QuadratureSpace::GetElementIntRule` which will return a quadrature for an element, not a face.

For now I think it is OK to remove this version of the `GetFaceQuadratureInterpolator` method. We can introduce it later when we add support for meshes with different types of faces and/or variable-order spaces where we want to vary the face quadratures.

v-dobrev commented on fem/fespace.hpp:
```diff
@@ -106,6 +108,11 @@ class FiniteElementSpace
+     * fes->GetMesh()->GetNBE() doesn't return the expected value in 3D
+     * because periodic meshes in 3D have some faces marked as boundary
+     * for vizualization purpose in GLvis. */
+   mutable int nbInteriorFaces, nbBoundaryFaces;
```
These are properties of the `Mesh` -- it makes sense to move them there.

YohannDudouit replied:
done.

v-dobrev commented on fem/pfespace.hpp:
```diff
@@ -264,6 +264,12 @@ class ParFiniteElementSpace : public FiniteElementSpace
+   virtual const Operator *GetFaceRestriction(ElementDofOrdering e_ordering,
+                                              FaceType type,
+                                              L2FaceValues mul = L2FaceValues::DoubleValued
+                                             ) const;
```
I assume this parallel version is different from the serial one because of the presence of shared faces. Can you add a sentence of two in the doxygen documentation about. For example, are shared faces treated as interior faces? If so, does the returned operator handle the communication needed to get the shared face values from neighbor processors?

YohannDudouit replied:
I used this comment to create documentation as you suggested.

v-dobrev commented on fem/pgridfunc.cpp:
```diff
@@ -130,7 +131,15 @@ void ParGridFunction::MakeRef(ParFiniteElementSpace *f, Vector &v, int v_offset)
+   else
+   {
+      *this = *tv;
+   }
```
This can now be simplified to just `prolong->Mult(*tv, *this);` -- the pointer cannot be `NULL` now.

v-dobrev commented on fem/plinearform.cpp:
```diff
@@ -33,13 +33,22 @@ void ParLinearForm::Update(ParFiniteElementSpace *pf, Vector &v, int v_offset)
+   {
+      tv = *this;
+   }
 }
```
Simplify to `prolong->MultTranspose(*this, tv);`.

v-dobrev commented on fem/plinearform.cpp:
```diff
@@ -33,13 +33,22 @@ void ParLinearForm::Update(ParFiniteElementSpace *pf, Vector &v, int v_offset)
    HypreParVector *tv = pfes->NewTrueDofVector();
-   pfes->GetProlongationMatrix()->MultTranspose(*this, *tv);
+   const Operator* prolong = pfes->GetProlongationMatrix();
+   if (!IsIdentityProlongation(prolong)) { prolong->MultTranspose(*this, *tv); }
```
This can be reverted too.

v-dobrev commented on fem/bilinearform_ext.cpp:
```diff
@@ -91,15 +134,10 @@ void PABilinearFormExtension::Update()
    FiniteElementSpace *fes = a->FESpace();
    height = width = fes->GetVSize();
-   trialFes = fes;
-   testFes = fes;
```
I think these two lines were removed by mistake.

YohannDudouit replied:
Reverted.

v-dobrev commented on fem/bilinearform_ext.hpp:
```diff
@@ -102,11 +102,16 @@ class PABilinearFormExtension : public BilinearFormExtension
+   mutable Vector faceBdrX, faceBdrY;
    const Operator *elem_restrict_lex; // Not owned
+   const Operator *int_face_restrict_lex; // Not owned
+   const Operator *bound_face_restrict_lex; // Not owned
```
Maybe change "bound" -> "bdr" here too.

YohannDudouit replied:
done.

v-dobrev commented on fem/bilininteg.cpp:
```diff
@@ -2617,6 +2633,14 @@ void DGTraceIntegrator::AssembleFaceMatrix(const FiniteElement &el1,
+const IntegrationRule &DGTraceIntegrator::GetRule(
+   Geometry::Type geom, int order, FaceElementTransformations &T)
+{
+   int int_order = T.Elem1->OrderW() + 2*order;
```
This is inconsistent with the formula in `AssembleFaceMatrix`. What was the reason to add `T.Elem1->OrderW()`?

v-dobrev commented on fem/bilininteg_dgtrace.cpp:
```diff
@@ -0,0 +1,1089 @@
+                                                  fes.GetMesh()->GetFaceBaseGeometry(0));
+   FaceElementTransformations &T = *fes.GetMesh()->GetFaceElementTransformations(
+                                      0);
+   const IntegrationRule *ir = &GetRule(el.GetGeomType(), el.GetOrder(), T);
```
If `this->IntRule` is not `NULL`, we should use that, e.g.:
```c++
    const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el.GetGeomType(), el.GetOrder(), T);
```

YohannDudouit replied:
done.

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+};
+
+/** @brief A class that performs interpolation from an E-vector to quadrature
+    point values and/or derivatives (Q-vectors) on the faces. */
```
It may be good to emphasize that the "E-vector" mentioned here is a "face E-vector".

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+    point values and/or derivatives (Q-vectors) on the faces. */
+/** An E-vector represents the element-wise discontinuous version of the FE
+    space and can be obtained, for example, from a GridFunction using the
+    Operator returned by FiniteElementSpace::GetElementRestriction().
```
Similarly, "An E-vector" -> "A face E-vector"; "element-wise" -> "face-wise"; "of the FE space" -> "of the trace FE space"; "`GetElementRestriction`" -> "`GetFaceRestriction`".

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+
+    The target quadrature points in the elements can be described either by an
+    IntegrationRule (all mesh elements must be of the same type in this case) or
+    by a QuadratureSpace. */
```
Similarly, "elements" -> "faces"; also, the `QuadratureSpace` is not quite suitable for describing face quadratures, so we should probably remove that part.

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+   friend class FiniteElementSpace; // Needs access to qspace and IntRule
+
+   const FiniteElementSpace *fespace;  ///< Not owned
+   const QuadratureSpace *qspace;      ///< Not owned
```
Remove?

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+                              const IntegrationRule &ir, FaceType type);
+
+   FaceQuadratureInterpolator(const FiniteElementSpace &fes,
+                              const QuadratureSpace &qs, FaceType type);
```
Remove.

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+   { use_tensor_products = !disable; }
+
+   /// Interpolate the E-vector @a e_vec to quadrature points.
+   /** The @a eval_flags are a bitwise mask of constants from the EvalFlags
```
`EvalFlags` -> `FaceEvalFlags`

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,222 @@
+      /** @brief Assuming the derivative at quadrature points form a matrix,
+          this flag can be used to compute and store their determinants. This
+          flag can only be used in Mult(). */
+      DETERMINANTS = 1 << 2,
```
`DETERMINANTS` is not quite accurate here (the Jacobians are not square), may call it `WEIGHTS`?

v-dobrev commented on fem/quadinterpolator.hpp:
```diff
@@ -0,0 +1,214 @@
+   /// Perform the transpose operation of Mult(). (TODO)
+   void MultTranspose(unsigned eval_flags, const Array<double> &W,
+                      const Vector &q_val,
+                      const Vector &q_der, Vector &e_vec) const;
```
I don't understand -- what does the parameter `W` do?

YohannDudouit replied:
I'll remove the whole function I think, is it okay with you?

v-dobrev replied:
Since it is not implemented yet, yes, I have no objection to remove it for now.

v-dobrev commented on fem/restriction.hpp:
```diff
@@ -0,0 +1,153 @@
+
+/// Operator that extracts Face degrees of freedom.
+/** Objects of this type are typically created and owned by FiniteElementSpace
+    objects, see FiniteElementSpace::GetElementRestriction(). */
```
`GetElementRestriction` -> `GetFaceRestriction`

v-dobrev commented on fem/restriction.hpp:
```diff
@@ -0,0 +1,153 @@
+
+/// Operator that extracts Face degrees of freedom.
+/** Objects of this type are typically created and owned by FiniteElementSpace
+    objects, see FiniteElementSpace::GetElementRestriction(). */
```
`GetElementRestriction` -> `GetFaceRestriction`

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
 };
 
+/** @brief Structure for storing mesh geometric factors: coordinates, Jacobians,
+    and determinants of the Jacobians. */
```
"mesh geometric factors" -> "face geometric factors"
"coordinates, Jacobians, and determinants of the Jacobians" -> "coordinates, Jacobians, determinants of the Jacobians, and normal vectors"

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
+/** @brief Structure for storing mesh geometric factors: coordinates, Jacobians,
+    and determinants of the Jacobians. */
+/** Typically objects of this type are constructed and owned by objects of class
+    Mesh. See Mesh::GetGeometricFactors(). */
```
`GetGeometricFactors` -> `GetFaceGeometricFactors`

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
+                        FaceType type);
+
+   /// Mapped (physical) coordinates of all quadrature points.
+   /** This array uses a column-major layout with dimensions (NQ x SDIM x NE)
```
`NE` -> `NF`

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
+
+   /// Jacobians of the element transformations at all quadrature points.
+   /** This array uses a column-major layout with dimensions (NQ x SDIM x DIM x
+       NE) where
```
`NE` -> `NF`

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
+   Vector J;
+
+   /// Determinants of the Jacobians at all quadrature points.
+   /** This array uses a column-major layout with dimensions (NQ x NE) where
```
`NE` -> `NF`

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1344,6 +1369,59 @@ class GeometricFactors
+   /** This array uses a column-major layout with dimensions (NQ x DIM x NE) where
+       - NQ = number of quadrature points per face, and
+       - DIM = Dimension of the mesh space
+       - NF = number of faces in the mesh. */
```
`NE` -> `NF`
`DIM` should be `SDIM` with `SDIM = space dimension of the mesh = mesh.SpaceDimension()`

v-dobrev commented on mesh/pmesh.cpp:
```diff
@@ -1712,6 +1712,28 @@ void ParMesh::DeleteFaceNbrData()
+   FiniteElementCollection* nfec;
+   if (discont)
+   {
+      const int type = 1; // Gauss-Lobatto points
```
Could use `type = BasisType::GaussLobatto` without a comment. 😄

v-dobrev commented on fem/fespace.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "../general/forall.hpp"
 #include "../mesh/mesh_headers.hpp"
 #include "fem.hpp"
+#include "quadinterpolator.hpp"
```
Make sure this header is included from `fem.hpp`.

v-dobrev commented on mesh/pmesh.cpp:
```diff
@@ -1712,6 +1712,27 @@ void ParMesh::DeleteFaceNbrData()
+   }
+   ParFiniteElementSpace* nfes = new ParFiniteElementSpace(this, nfec, space_dim,
+                                                           ordering);
+   SetNodalFESpace(nfes);
```
I think this line can be removed.

YohannDudouit replied:
done.

v-dobrev commented:
Merged in `next` for testing...

v-dobrev commented:
At the moment, merging in `next` has a lot of conflics due to the changes in #1204 which are already in `next` (but not in `master`). Let's wait for #1204 to be merged in `master`, then merge `master` here, and then merge this PR into `next`.

tzanio commented:
Please resolve the merge conflict (after merging #1204 in master)

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Several issues from the regression tests tonight:

1.  The following runs failed in [sample-runs-baseline-tux375.diff](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/1d0dd8fa448b5b52fafe344dc13429df4b7a490e#tux375/2020-03-03-llnl-gh-next/sample-runs-baseline-tux375.diff):

```
ex9 -pa -m ../data/periodic-cube.mesh -no-vis
mpirun -np 4 ex9p -pa -m ../data/periodic-cube.mesh -no-vis
```

2. Memory leaks reported in [memcheck.err](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/1d0dd8fa448b5b52fafe344dc13429df4b7a490e#tux375/2020-03-03-llnl-gh-next/memcheck.err), e.g.

```
 90,520 (48 direct, 90,472 indirect) bytes in 1 blocks are definitely lost in loss record 47 of 50
    at 0x4C2A483: operator new(unsigned long) (vg_replace_malloc.c:344)
    by 0x442379: mfem::RAP(mfem::HypreParMatrix const*, mfem::HypreParMatrix const*) (hypre.cpp:1653)
    by 0x47262A: mfem::OperatorHandle::MakePtAP(mfem::OperatorHandle&, mfem::OperatorHandle&) (handle.cpp:146)
    by 0x6B2111: mfem::ParBilinearForm::ParallelAssemble(mfem::OperatorHandle&, mfem::SparseMatrix*) (pbilinearform.cpp:179)
    by 0x6B221A: mfem::ParBilinearForm::ParallelAssemble(mfem::SparseMatrix*) (pbilinearform.cpp:185)
    by 0x40E670: mfem::ParBilinearForm::ParallelAssemble() (pbilinearform.hpp:94)
    by 0x40B3EE: FE_Evolution::FE_Evolution(mfem::ParBilinearForm&, mfem::ParBilinearForm&, mfem::Vector const&) (ex9p.cpp:512)
    by 0x40A9A6: main (ex9p.cpp:430)

 437,452 (48 direct, 437,404 indirect) bytes in 1 blocks are definitely lost in loss record 50 of 50
    at 0x4C2A483: operator new(unsigned long) (vg_replace_malloc.c:344)
    by 0x442379: mfem::RAP(mfem::HypreParMatrix const*, mfem::HypreParMatrix const*) (hypre.cpp:1653)
    by 0x47262A: mfem::OperatorHandle::MakePtAP(mfem::OperatorHandle&, mfem::OperatorHandle&) (handle.cpp:146)
    by 0x6B2111: mfem::ParBilinearForm::ParallelAssemble(mfem::OperatorHandle&, mfem::SparseMatrix*) (pbilinearform.cpp:179)
    by 0x6B221A: mfem::ParBilinearForm::ParallelAssemble(mfem::SparseMatrix*) (pbilinearform.cpp:185)
    by 0x40E670: mfem::ParBilinearForm::ParallelAssemble() (pbilinearform.hpp:94)
    by 0x40B418: FE_Evolution::FE_Evolution(mfem::ParBilinearForm&, mfem::ParBilinearForm&, mfem::Vector const&) (ex9p.cpp:515)
    by 0x40A9A6: main (ex9p.cpp:430)
```

```

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented on examples/ex9.cpp:
```diff
@@ -191,6 +201,9 @@ int main(int argc, char *argv[])
 
+   Device device(device_config);
+   device.Print();
+
```
It is really awesome to see the Device support spread through the examples! (So far in ex1, ex1p, ex3, ex3p, ex6, ex6p, ex9, ex9p, ex24 and ex24p)

v-dobrev commented on fem/pgridfunc.cpp:
```diff
@@ -225,23 +228,27 @@ void ParGridFunction::ExchangeFaceNbrData()
 
-   const double *h_data = this->HostRead();
-   for (int i = 0; i < send_data.Size(); i++)
+   auto d_data = this->Read();
```
I think this implementation can be improved to handle better the case when gpu is enabled but `this->data` is on host when this method is called. In that case, it will be better to do the buffer copies on host avoiding the copy of the whole data to the gpu.

There's no need to change the implementation now, I just wanted to point this out for future considerations.
